src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,144 @@
+                    + index + " passed to method is invalid");
+        }
+        if (size == index) {
+            add(value);
+        } else {
+            Node<T> currentNode = findNodeByIndex(index);
+            Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+            currentNode.prev = newNode;
+            if (index == 0) {
+                head = newNode;
+            } else {
+                newNode.prev.next = newNode;
+            }
+            size++;
+        }
+
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexValidation(index);
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexValidation(index);
+        Node<T> currentNode = findNodeByIndex(index);
+        T oldValue = currentNode.item;
+        currentNode.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexValidation(index);
+        Node<T> node = findNodeByIndex(index);
+        if (head == tail) {
+            head = null;
+            tail = null;
+        } else if (index == 0) {
+            head.next.prev = null;
+            head = head.next;
+        } else if (index == size - 1) {
+            tail.prev.next = null;
+            tail = tail.prev;
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+        }
+        size--;
+        return node.item;
     }
 
     @Override
     public boolean remove(T object) {
+        int i = 0;
+        if (object == null) {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (x.item == object) {
+                    remove(i);
+                    return true;
+                }
+                i++;
+            }
+        } else {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (x.item.equals(object)) {
+                    remove(i);
+                    return true;
+                }
+                i++;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndexValidation(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("The index "
+                    + index + " passed to method is invalid");
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
```

Comment:
if the index is in the first part of list, we can begin the movement from the head, otherwise from the tail
it will improve performance